python/GafferImageUI/CatalogueUI.py

```diff
@@ -131,6 +135,81 @@
 
 )
 
+
+##########################################################################
+# Viewer hot-keys
+##########################################################################
+
+def addCatalogueHotkeys( editor ) :
+
+	if not isinstance( editor, GafferUI.Viewer ) :
+		return
+
+	editor.keyPressSignal().connect( _handleCatalogueHotkeys, scoped = False )
+
+def _handleCatalogueHotkeys( viewer, event ) :
+
+	# Up/Down arrows need to walk upstream of the viewer input and look for
+	# a Catalogue node and increment/decrement its active index
+
+	if event.key not in ( "Down", "Up" ) :
+		return False
+
+	if not isinstance( viewer.view(), GafferImageUI.ImageView ) :
+		return False
+
+	catalogue = __findCatalogue( viewer.view() )
+	if catalogue is None :
+		return False
+
+	__incrementImageIndex( catalogue, event.key )
+
+	return True
+
+def __findCatalogue( node ) :
+
+	if isinstance( node, GafferImage.Catalogue ) :
+		return node
+	else :
+		for inPlug in GafferImage.ImagePlug.InputRange( node ) :
+			upstreamPlug = inPlug.getInput()
```

Comment:
If we used `inPlug.source()` instead of `inPlug.getInput()` and `upstreamPlug != inPlug` instead of `upstreamPlug is None`, we'd avoid a bunch of work when traversing through pass-through nodes like Dot.

Reply:
Improved in https://github.com/GafferHQ/gaffer/pull/3387/commits/33ac1996f8c09b904eb9fbebcbcbec8772b5c489

Reply:
`source()` never returns `None`, so the `upstreamPlug is None` test remaining below is meaningless. As per my original comment, the right test is `upstreamPlug != inPlug`.

Reply:
Sorry of course, sorted in the squash. https://github.com/GafferHQ/gaffer/blob/7db330c669c22049fc7536be3f29897c95e90ad7/python/GafferImageUI/CatalogueUI.py#L177